File: backend/search/score_test.go
package search

import (
	"os"
	"path/filepath"
	"testing"
	"time"
)

// mkdirAllFile 建出中间目录并落一个空文件。
func mkdirAllFile(p string) error {
	if err := os.MkdirAll(filepath.Dir(p), 0o755); err != nil {
		return err
	}
	return os.WriteFile(p, nil, 0o644)
}

func TestScoreSubsequenceAndMiss(t *testing.T) {
	if _, ok := Score("ovw", "Overview.tsx"); !ok {
		t.Fatal("首字母缩写 ovw 应能命中 Overview.tsx")
	}
	if _, ok := Score("zzz", "Overview.tsx"); ok {
		t.Fatal("不存在的子序列不应命中")
	}
	if _, ok := Score("", "anything"); ok {
		t.Fatal("空查询不应命中")
	}
}

func TestScorePrefersWordBoundary(t *testing.T) {
	// 同样是子序列命中，词首（路径段开头）要压过散落在中间的巧合
	head, ok1 := Score("api", "backend/api/api.go")
	tail, ok2 := Score("api", "backend/xapixzz/notes.md")
	if !ok1 || !ok2 {
		t.Fatal("两个目标都应命中")
	}
	if head.Score <= tail.Score {
		t.Fatalf("词首命中应更高：%d vs %d", head.Score, tail.Score)
	}
}

func TestScorePrefersConsecutive(t *testing.T) {
	solid, _ := Score("search", "search.go")
	spread, _ := Score("search", "s_e_a_r_c_h.go")
	if solid.Score <= spread.Score {
		t.Fatalf("连续命中应更高：%d vs %d", solid.Score, spread.Score)
	}
}

func TestScoreCaseInsensitive(t *testing.T) {
	a, ok := Score("OVERVIEW", "overview.tsx")
	if !ok || a.Score <= 0 {
		t.Fatal("大小写不应影响命中")
	}
}

func TestScoreMultiTermIsAnd(t *testing.T) {
	if _, ok := Score("api search", "backend/api/search.go"); !ok {
		t.Fatal("两个词都在目标里应命中")
	}
	if _, ok := Score("api zzz", "backend/api/search.go"); ok {
		t.Fatal("有一个词不在目标里就不该命中")
	}
}

func TestScorePositionsAreValid(t *testing.T) {
	m, ok := Score("apgo", "backend/api/api.go")
	if !ok {
		t.Fatal("应命中")
	}
	if len(m.Positions) != 4 {
		t.Fatalf("位置数应等于查询字符数，得到 %d", len(m.Positions))
	}
	target := []rune("backend/api/api.go")
	want := []rune("apgo")
	for i, p := range m.Positions {
		if p < 0 || p >= len(target) {
			t.Fatalf("位置越界：%d", p)
		}
		if i > 0 && p <= m.Positions[i-1] {
			t.Fatalf("位置应严格递增：%v", m.Positions)
		}
		if target[p] != want[i] {
			t.Fatalf("位置 %d 指向 %q，期望 %q", p, string(target[p]), string(want[i]))
		}
	}
}

func TestScoreCJK(t *testing.T) {
	if _, ok := Score("概览", "概览页 Overview"); !ok {
		t.Fatal("中文子串应命中")
	}
	if _, ok := Score("概设", "概览设计"); !ok {
		t.Fatal("中文子序列应命中")
	}
}

func TestBestPrefersPrimaryField(t *testing.T) {
	// 名字命中打满分，路径命中打折：真叫 search.go 的那个要排在「路径里带 search」前面
	name, idx, ok := Best("search", "search.go", "/home/x/deep/nested/dir/search/other.go")
	if !ok || idx != 0 {
		t.Fatalf("应命中主字段，得到 idx=%d ok=%v", idx, ok)
	}
	sub, idx2, ok2 := Best("search", "other.go", "/home/x/search/other.go")
	if !ok2 || idx2 != 1 {
		t.Fatalf("主字段不匹配时应回落到副字段，得到 idx=%d ok=%v", idx2, ok2)
	}
	if sub.Score >= name.Score {
		t.Fatalf("副字段命中应低于主字段：%d vs %d", sub.Score, name.Score)
	}
}

func TestFileIndexWalk(t *testing.T) {
	root := t.TempDir()
	mk := func(rel string) {
		p := filepath.Join(root, rel)
		if err := mkdirAllFile(p); err != nil {
			t.Fatal(err)
		}
	}
	mk("src/app.tsx")
	mk("src/lib/util.go")
	mk("node_modules/pkg/index.js")
	mk(".git/config")
	mk("README.md")

	ix := NewFileIndex()
	files, truncated := ix.Rebuild(root) // 同步版：测遍历规则本身，不测后台调度
	if truncated {
		t.Fatal("小目录不该被截断")
	}
	got := map[string]bool{}
	for _, f := range files {
		got[filepath.ToSlash(f)] = true
	}
	for _, want := range []string{"src/app.tsx", "src/lib/util.go", "README.md"} {
		if !got[want] {
			t.Fatalf("清单缺 %s：%v", want, files)
		}
	}
	for _, skip := range []string{"node_modules/pkg/index.js", ".git/config"} {
		if got[skip] {
			t.Fatalf("%s 不该进索引", skip)
		}
	}
}

func TestFileIndexTruncates(t *testing.T) {
	root := t.TempDir()
	for i := 0; i < 20; i++ {
		if err := mkdirAllFile(filepath.Join(root, "f"+string(rune('a'+i))+".txt")); err != nil {
			t.Fatal(err)
		}
	}
	ix := NewFileIndex()
	ix.MaxFiles = 5
	files, truncated := ix.Rebuild(root)
	if !truncated {
		t.Fatal("超过上限应标记截断")
	}
	if len(files) > 5 {
		t.Fatalf("收录数不应超过上限，得到 %d", len(files))
	}
}

func TestFileIndexBuildsInBackground(t *testing.T) {
	root := t.TempDir()
	if err := mkdirAllFile(filepath.Join(root, "a.txt")); err != nil {
		t.Fatal(err)
	}
	ix := NewFileIndex()
	// 第一次问：立刻返回，清单还没建好
	if files, _, ready := ix.Files(root); ready || len(files) != 0 {
		t.Fatalf("首次调用应立刻返回空清单且 ready=false，得到 files=%v ready=%v", files, ready)
	}
	deadline := time.Now().Add(3 * time.Second)
	for time.Now().Before(deadline) {
		if files, _, ready := ix.Files(root); ready {
			if len(files) != 1 {
				t.Fatalf("建好后应有 1 个文件，得到 %v", files)
			}
			return
		}
		time.Sleep(10 * time.Millisecond)
	}
	t.Fatal("后台索引没在 3 秒内建好")
}

File: frontend/src/search/score.test.ts
// 与 backend/search/score_test.go 同一批用例：两份实现只要有一份漂了，
// 合并后的排序就会自相矛盾（同一个东西在两路里分数不同）。
import { describe, expect, it } from 'vitest'
import { fuzzyBest, fuzzyScore } from './score'
import { rankLocal, dedupeByKey, sortByKind } from './rank'

describe('fuzzyScore', () => {
  it('首字母缩写能命中，不存在的子序列不命中', () => {
    expect(fuzzyScore('ovw', 'Overview.tsx')).not.toBeNull()
    expect(fuzzyScore('zzz', 'Overview.tsx')).toBeNull()
    expect(fuzzyScore('', 'anything')).toBeNull()
  })

  it('词首命中压过散落在中间的巧合', () => {
    const head = fuzzyScore('api', 'backend/api/api.go')!
    const tail = fuzzyScore('api', 'backend/xapixzz/notes.md')!
    expect(head.score).toBeGreaterThan(tail.score)
  })

  it('连续命中压过被打散的命中', () => {
    const solid = fuzzyScore('search', 'search.go')!
    const spread = fuzzyScore('search', 's_e_a_r_c_h.go')!
    expect(solid.score).toBeGreaterThan(spread.score)
  })

  it('忽略大小写', () => {
    expect(fuzzyScore('OVERVIEW', 'overview.tsx')).not.toBeNull()
  })

  it('多个词是 AND', () => {
    expect(fuzzyScore('api search', 'backend/api/search.go')).not.toBeNull()
    expect(fuzzyScore('api zzz', 'backend/api/search.go')).toBeNull()
  })

  it('命中位置严格递增且指向对的字符', () => {
    const m = fuzzyScore('apgo', 'backend/api/api.go')!
    const target = Array.from('backend/api/api.go')
    const want = Array.from('apgo')
    expect(m.positions).toHaveLength(4)
    m.positions.forEach((p, i) => {
      expect(target[p]).toBe(want[i])
      if (i > 0) expect(p).toBeGreaterThan(m.positions[i - 1])
    })
  })

  it('中文按子串/子序列匹配', () => {
    expect(fuzzyScore('概览', '概览页 Overview')).not.toBeNull()
    expect(fuzzyScore('概设', '概览设计')).not.toBeNull()
  })
})

describe('fuzzyBest', () => {
  it('主字段命中给位置，副字段命中打折且不给位置', () => {
    const name = fuzzyBest('search', ['search.go', '/home/x/deep/search/other.go'])!
    expect(name.positions).toBeDefined()
    const sub = fuzzyBest('search', ['other.go', '/home/x/search/other.go'])!
    expect(sub.positions).toBeUndefined()
    expect(sub.score).toBeLessThan(name.score)
  })

  it('一个字段都不匹配就返回 null', () => {
    expect(fuzzyBest('zzz', ['a.go', 'b/c.go'])).toBeNull()
  })
})

describe('rankLocal', () => {
  it('丢掉不匹配的，按分数排', () => {
    const items = [
      { key: 'a', title: 'Overview' },
      { key: 'b', title: 'Projects' },
      { key: 'c', title: 'Over the top view' },
    ]
    const out = rankLocal('overview', items)
    expect(out.map((i) => i.key)).toEqual(['a', 'c'])
  })

  it('别名（keywords）也参与匹配', () => {
    const out = rankLocal('files', [{ key: 'a', title: '文件', keywords: 'files' }])
    expect(out).toHaveLength(1)
  })
})

describe('合并规则', () => {
  it('同 key 只留第一条', () => {
    const rows = [{ key: 'x', n: 1 }, { key: 'x', n: 2 }, { key: 'y', n: 3 }]
    expect(dedupeByKey(rows).map((r) => r.n)).toEqual([1, 3])
  })

  it('先按类别固定顺序，再按分数', () => {
    const rows = [
      { key: '1', kind: 'file' as const, score: 999 },
      { key: '2', kind: 'page' as const, score: 1 },
      { key: '3', kind: 'session' as const, score: 5 },
      { key: '4', kind: 'session' as const, score: 50 },
    ]
    expect(sortByKind(rows).map((r) => r.key)).toEqual(['2', '4', '3', '1'])
  })
})
